src/Amalgam/SeparableBoxFilterDataStore.h

```diff
@@ -144,28 +144,22 @@ class SeparableBoxFilterDataStore
 		//if big enough (enough entities and/or enough columns), try to use multithreading
 		if(num_columns_added > 1 && (numEntities > 10000 || (numEntities > 200 && num_columns_added > 10)))
 		{
-			auto enqueue_task_lock = Concurrency::threadPool.BeginEnqueueBatchTask();
-			if(enqueue_task_lock.AreThreadsAvailable())
-			{
-				std::vector<std::future<void>> columns_completed;
-				columns_completed.reserve(num_columns);
-
-				for(size_t i = num_previous_columns; i < num_columns; i++)
-				{
-					columns_completed.emplace_back(
-						Concurrency::threadPool.EnqueueBatchTask([this, &entities, i]() { BuildLabel(i, entities); })
-					);
-				}
+			std::vector<std::future<void>> columns_completed;
+			columns_completed.reserve(num_columns);
 
-				enqueue_task_lock.Unlock();
+			for(size_t i = num_previous_columns; i < num_columns; i++)
+			{
+				columns_completed.emplace_back(
+					Concurrency::urgentThreadPool.EnqueueSingleTask([this, &entities, i]() { BuildLabel(i, entities); })
+				);
+			}
 
-				Concurrency::threadPool.ChangeCurrentThreadStateFromActiveToWaiting();
-				for(auto &future : columns_completed)
-					future.wait();
-				Concurrency::threadPool.ChangeCurrentThreadStateFromWaitingToActive();
+			Concurrency::urgentThreadPool.ChangeCurrentThreadStateFromActiveToWaiting();
```

Comment:
Is the current thread guaranteed to be in the urgentThreadPool?

Reply:
I suppose it doesn't actually matter if it is an urgent thread or not but still... seems like we are changing the urgentpool thread state but it could be from a non-urgentpool thread.